ref/go/src/bech32/bech32_test.go

```diff
@@ -0,0 +1,273 @@
+		},
+	},
+	item{"bc1zw508d6qejxtdg4y5r3zarvaryvg6kdaj",
+		[]int{
+			0x82, 0x10, 0x75, 0x1e, 0x76, 0xe8, 0x19, 0x91, 0x96, 0xd4, 0x54,
+			0x94, 0x1c, 0x45, 0xd1, 0xb3, 0xa3, 0x23,
+		},
+	},
+	item{"tb1qqqqqp399et2xygdj5xreqhjjvcmzhxw4aywxecjdzew6hylgvsesrxh6hy",
+		[]int{
+			0x00, 0x20, 0x00, 0x00, 0x00, 0xc4, 0xa5, 0xca, 0xd4, 0x62, 0x21,
+			0xb2, 0xa1, 0x87, 0x90, 0x5e, 0x52, 0x66, 0x36, 0x2b, 0x99, 0xd5,
+			0xe9, 0x1c, 0x6c, 0xe2, 0x4d, 0x16, 0x5d, 0xab, 0x93, 0xe8, 0x64,
+			0x33,
+		},
+	},
+}
+
+var INVALIDADDRESS = []string{
+	"tc1qw508d6qejxtdg4y5r3zarvary0c5xw7kg3g4ty",
+	"bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t5",
+	"BC13W508D6QEJXTDG4Y5R3ZARVARY0C5XW7KN40WF2",
+	"bc1rw5uspcuh",
+	"bc10w508d6qejxtdg4y5r3zarvary0c5xw7kw508d6qejxtdg4y5r3zarvary0c5xw7kw5rljs90",
+	"BC1QR508D6QEJXTDG4Y5R3ZARVARYV98GJ9P",
+	"tb1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3q0sL5k7",
+	"tb1pw508d6qejxtdg4y5r3zarqfsj6c3",
+	"tb1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3pjxtptv",
+}
+
+func TestValidChecksum(t *testing.T) {
+	for _, test := range VALIDCHECKSUM {
+		hrp, data, err := bech32.Decode(test)
+		if err != nil {
+			t.Errorf("Valid checksum for %s : FAIL / error %+v\n", test, err)
+		} else {
+			t.Logf("Valid checksum for %s : ok / hrp : %+v , data : %+v\n", test, hrp, data)
+		}
+	}
+}
+
+func TestValidAddress(t *testing.T) {
+	for _, test := range VALIDADDRESS {
+		hrp := "bc"
+		version, program, err := bech32.SegwitAddrDecode(hrp, test.address)
+		if err != nil {
+			hrp = "tb"
+			version, program, err = bech32.SegwitAddrDecode(hrp, test.address)
+		}
+		ok := err == nil
+		if ok {
+			output := segwitScriptpubkey(version, program)
+			ok = reflect.DeepEqual(output, test.scriptpubkey)
+		}
+		if ok {
+			recreate, err := bech32.SegwitAddrEncode(hrp, version, program)
+			if err == nil {
+				ok = recreate == strings.ToLower(test.address)
+			}
+		}
+		if ok {
+			t.Logf("Valid address %v : ok\n", test.address)
+		} else {
+			t.Errorf("Valid address %v : FAIL\n", test.address)
+		}
+	}
+}
+
+func TestInvalidAddress(t *testing.T) {
+	for _, test := range INVALIDADDRESS {
+		ok := false
+		_, _, err := bech32.SegwitAddrDecode("bc", test)
+		if err != nil {
+			t.Logf("%v : %v\n", test, err)
+			_, _, err = bech32.SegwitAddrDecode("tb", test)
+			if err != nil {
+				t.Logf("%v : %v\n", test, err)
+				ok = true
+			}
+		}
+		if ok {
+			t.Logf("Invalid address %v : ok\n", test)
+		} else {
+			t.Errorf("Invalid address %v : FAIL\n", test)
+		}
+	}
+}
+
+// add coverage tests
+
+func TestCoverage(t *testing.T) {
+	var err error
+	_, err = bech32.SegwitAddrEncode("bc", 0, []int{-1})
+	if err == nil {
+		t.Errorf("Coverage SegwitAddrEncode invalid data range error case : FAIL")
+	} else {
+		t.Log("Coverage SegwitAddrEncode invalid data range error case : ok / error :", err)
+	}
+	_, err = bech32.SegwitAddrEncode("bc", 33, []int{})
+	if err == nil {
+		t.Errorf("Coverage SegwitAddrEncode invalid data error case : FAIL")
```

Comment:
this is checking the invalid version 33, not the invalid data, right? If happens to catch that line because you check the version after encoding it as data, but that won't be the case if you add the version check in SegwitAddrEncode().

You can just move the 33 inside the int slice and use a valid version.